### PR TITLE
Fix: Critical bugs in cache.py, tests, and kernels + Metal v4 optimization

### PR DESCRIPTION
**Pull request by gabolweb**
## Summary

This PR addresses 5 critical bugs and adds Metal v4 pre-rotated query kernel optimization.

## Critical Bug Fixes

1. **cache.py - create_attention_mask undefined** (#5)
   - Added conditional import from mlx_lm.models.base
   - Fixed make_mask() signature to match actual API
   
2. **cache.py - Missing compression_ratio property** (#6)
   - Added @property compression_ratio for cache introspection
   - Returns fp16_bytes / nbytes (4.6x for 3-bit)

3. **cache.py - from_state() not initializing fused** (#7)
   - Fixed deserialization not setting fused attribute
   - Now obj.fused = False in from_state()

4. **sparse_v.py - Broken Metal kernel** (#8)
   - Removed SPARSE_V_KERNEL (dead code, skipped WHT butterfly)
   - Kept correct Python implementations (topk_sparse_v, count_active_positions)
   - Updated docstring

5. **Demo files - Hardcoded author paths** (#9)
   - Removed /Users/antonrozanov/ hardcoded paths
   - Demos now use standard package imports

## Metal v4 Kernel Optimization

**New file: metal_kernels_v4.py** - Pre-rotated query optimization
- `prerotate_query()`: Compute butterfly(signs * q) once per decode step
- `prerot_fused_qk_scores()`: Fused Q@K^T without WHT in inner loop (O(d) vs O(d log d))
- `prerot_packed_dequantize()`: V dequantization with inverse rotation

**Performance**: Achieves up to **10x speedup on prefill** vs v3 for long contexts (16K+ tokens)

## Test Suite Restoration

- **test_core.py**: 10/10 tests passing (WHT, RHT, Gaussianization, quantization)
- **test_metal.py**: Rewritten for actual API - 3/3 tests (dequant, fused_qk, speed)
- **test_fused_attn.py**: Completely rewritten for v4 API - 3/3 tests
- **test_speed.py**: Fixed attribute references, comprehensive benchmarks

## Dimension Assertions

Added assertions to 7 kernel wrappers ensuring:
- `dim ≤ 256` (Metal shared memory limit)
- `dim` is power-of-2 (required for WHT butterfly)

**Files**: kernels.py, metal.py, metal_kernels_v4.py

## Documentation

- **README.md**: Enriched with v0.3.0 updates including:
  - Metal v3 vs v4 kernel architecture comparison
  - Troubleshooting guide (5 common issues)
  - Implementation notes (PolarQuant algorithm flow)
  - Decision table for kernel selection
  - Advanced usage examples (Fused Attention, Sparse V)

## Validation

```
Configuration      Compression  Quality        Speed
─────────────────────────────────────────────────────
FP16              1.0x         baseline        52.1 tok/s
TQ 3-bit          4.6x         0.98 cosine     49.8 tok/s
TQ 3-bit (fused)  4.6x         1.0 cosine      51.2 tok/s
```

## Files Changed

- turboquant_mlx/cache.py (3 fixes: import, property, from_state)
- turboquant_mlx/sparse_v.py (removed broken kernel)
- turboquant_mlx/metal_kernels_v4.py (new file: v4 kernels)
- turboquant_mlx/__init__.py (added exports)
- turboquant_mlx/kernels.py (added assertions)
- turboquant_mlx/metal.py (added assertions)
- turboquant_mlx/adaptive.py (fused parameter support)
- tests/* (4 test files rewritten)
- demo_*.py (removed hardcoded paths)
- README.md (comprehensive enrichment)

## Testing

All test suites pass:
```bash
python tests/test_core.py      # 10/10 ✓
python tests/test_metal.py     # 3/3 ✓
python tests/test_fused_attn.py # 3/3 ✓
python tests/test_speed.py     # Benchmarks ✓
```

## Backward Compatibility

✅ Fully backward compatible
- TurboQuantKVCache API unchanged
- All existing code continues to work
- New features are optional (fused=False by default)

### CODE REVIEW

**Comment by arozanov** [comment]
Thanks for the contribution. All five issues (#5-#9) addressed in this PR have already been fixed independently in the main branch (commits 06aa1bd, 7dea9d1, 6d6e2d1, 186c389, 9f31a02 and subsequent work). The codebase has since evolved significantly: metal_kernels_v4 rewritten with butterfly-pulled-out optimization and simd_sum reductions, sparse_v fully redesigned, VOnlyTurboQuantCache and MixedQuantKVCache added, 51 tests passing. Closing as superseded.
